File: python/euler.py
import time
import math
import pickle
import itertools
import subprocess32
import requests

class TimeWrapper(object):
    """
    Times the code running inside a with, then prints when exiting the with.
    """
    def __enter__(self):
        self._time_started = time.clock()
    def __exit__(self, type, value, traceback):
        time_passed = time.clock() - self._time_started
        print "{0:.6f}".format(time_passed)
        return False

class PrimesRepeat(object):
    """
    Generates primes in such a way that it can be called multiple
    times with different values and only generate new primes when
    needed.
    """
    def __init__(self):
        self.l = [2]
        self.last = 2
        
    def is_prime(self,n):
        if n == self.last:
            return True
        elif n < self.last:
            return n in self.l
        else:
            for i in itertools.count(self.last+1):
                if is_prime(i):
                    self.l.append(i)
                    #End the loop if i >= n
                    if i == n:
                        self.last = i
                        return True
                    elif i > n:
                        self.last = i
                        return False
                    
    def get_primes_until(self,n):
        self.is_prime(n) #Generate primes until past n if needed
        for i in self.l:
            #End the loop before i gets >= n
            if i >= n:
                return
            yield i

def factorize(n):
    """
    Returns a tuple with all the prime factors of a given number.
    """
    call_output = subprocess32.check_output(['factor',str(n)])
    number,factors = [part.strip() for part in call_output.split(':')]
    if not factors:
        return ()
    factors_tuple = tuple([int(number.strip()) for number in factors.split(' ')])
    return factors_tuple

def is_prime(n):
    """
    Return whether a single number is prime.
    """
    limit = int( math.ceil( math.sqrt(n) ) )
    for i in xrange(2,limit+1):
        if n % i == 0:
            return False
    return True

def gcd(a,b):
    """
    Return the greatest common divisor of two numbers.
    """
    if b == 0:
        return a
    else:
        return gcd(b,a%b)

def primes_sieve(limit):
    """
    Generate primes using a sieve, from 2 to a given limit.
    """
    sieve = [True]*limit
    sieve[0] = False
    sieve[1] = False
    for i in xrange(2,limit):
        if sieve[i]:
            yield i
            for j in xrange(i*2, limit, i):
                sieve[j] = False

def read_url(url):
    """
    Return the raw text from an url.
    """
    r = requests.get(url)
    return r.text

def sum_divisors(n):
    """
    Return the sum of the porper divisors of n
    """
    factors = factorize(n)

    result = 1
    for p in set(factors):
        a = factors.count(p)
        result *= sum([p**ai for ai in range(a+1)])
    
    return result-n

File: csharp/euler.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Collections.Generic;

namespace Euler
{
  /// <summary>
  /// Generates primes in such a way that it can be called multiple times with
  /// different values and only generate new primes when needed.
  /// </summary>
  public class PrimesRepeat
  {
    private List<long> primes;
    private long last;

    public PrimesRepeat()
    {
      this.primes = new List<long> { 2 };
      this.last = 2;
    }

    public bool IsPrime(long n)
    {
      if (n == this.last)
      {
        return true;
      }
      else if (n < this.last)
      {
        return this.primes.BinarySearch(n) >= 0;
      }
      else
      {
        for (long i = this.last+1;; i++)
        {
          if (Euler.IsPrime(i))
          {
            this.primes.Add(i);
            if (i == n)
            {
              this.last = i;
              return true;
            }
            else if (i > n)
            {
              this.last = i;
              return false;
            }
          }
        }
      }
    }
  }

  public static class Euler
  {
    /// <summary>
    /// Execute and time the given action. Display the time taken to run.
    /// </summary>
    public static void TimeWrapper(Action action)
    {
      var timer = new Stopwatch();
      timer.Start();
      action();
      timer.Stop();
      Console.WriteLine(timer.Elapsed.TotalSeconds);
    }

    /// <summary>
    /// Return the prime factors of the given number.
    /// </summary>
    public static long[] Factorize(long n)
    {
      string output;
      using (var p = new Process())
      {
        p.StartInfo.FileName = "factor";
        p.StartInfo.Arguments = n.ToString();
        p.StartInfo.UseShellExecute = false;
        p.StartInfo.RedirectStandardOutput = true;
        p.Start();

        output = p.StandardOutput.ReadToEnd();
        p.WaitForExit();
      }

      var numberAndFactors = output.Split(':');
      if (string.IsNullOrWhiteSpace(numberAndFactors[1]))
        return new long[] { };

      return numberAndFactors[1].Trim().Split()
        .Select(x => long.Parse(x.Trim())).ToArray();
    }

    /// <summary>
    /// Return whether a number is prime.
    /// </summary>
    public static bool IsPrime(long n)
    {
      var limit = Math.Ceiling(Math.Sqrt(n));
      for (long divisor = 2; divisor <= limit; divisor++)
      {
        if (n % divisor == 0)
          return false;
      }
      return true;
    }

    /// <summary>
    /// Return the greatest common divisor of two numbers.
    /// </summary>
    public static long Gcd(long a, long b)
    {
      if (b == 0)
        return a;
      else
        return Gcd(b, a%b);
    }

    /// <summary>
    /// Return an iterator of all the primes up to the given limit.
    /// </summary>
    public static IEnumerable<long> PrimesSieve(long limit)
    {
      if (limit < 2)
      {
        yield break;
      }

      var sieve = new bool[limit];
      sieve[0] = false;
      sieve[1] = false;
      for (int i=2; i<limit; i++)
      {
        sieve[i] = true;
      }

      for (int i=2; i<limit; i++)
      {
        if (sieve[i])
        {
          yield return i;
          for (int j=i*2; j<limit; j+=i)
          {
            sieve[j] = false;
          }
        }
      }
    }

    /// <summary>
    /// Return the text from an url
    /// </summary>
    public static string ReadURL(string address)
    {
      var client = new System.Net.WebClient();
      string reply = client.DownloadString (address);

      Console.WriteLine (reply);
      return null;
    }

    /// <summary>
    /// Return the sum of the propper divisors of n
    /// </summary>
    public static long SumDivisors(long n)
    {
      var factors = Euler.Factorize(n);
      long result = 1;
      foreach (var p in factors.Distinct()) {
        var a = factors.Where(x => x == p).Count();
        result *= Enumerable.Range(0, a+1).Select(ai => (long)Math.Pow(p, ai)).Sum();
      }
      return result - n;
    }
  }
}
